File: gitdot-web/packages/api/src/endpoint/organization/list-organization-members.ts
import { z } from "zod";
import { OrganizationMemberResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const ListOrganizationMembersRequest = z.object({
  role: z.string().optional(),
});
export type ListOrganizationMembersRequest = z.infer<
  typeof ListOrganizationMembersRequest
>;

export const ListOrganizationMembersResponse = z.array(
  OrganizationMemberResource,
);
export type ListOrganizationMembersResponse = z.infer<
  typeof ListOrganizationMembersResponse
>;

export const ListOrganizationMembers = {
  path: "/organization/{org_name}/members",
  method: "GET",
  request: ListOrganizationMembersRequest,
  response: ListOrganizationMembersResponse,
} as const satisfies Endpoint;
export type ListOrganizationMembers = typeof ListOrganizationMembers;

File: gitdot-api/src/endpoint/organization/list_organization_members.rs
use serde::{Deserialize, Serialize};

use crate::{
    endpoint::Endpoint,
    resource::{OrganizationMemberResource, common::Page},
};

pub struct ListOrganizationMembers;

impl Endpoint for ListOrganizationMembers {
    const PATH: &'static str = "/organization/{org_name}/members";
    const METHOD: http::Method = http::Method::GET;

    type Request = ListOrganizationMembersRequest;
    type Response = ListOrganizationMembersResponse;
}

#[derive(ApiRequest, Debug, Default, Serialize, Deserialize)]
pub struct ListOrganizationMembersRequest {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub role: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cursor: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
}

pub type ListOrganizationMembersResponse = Page<OrganizationMemberResource>;
